tracing: add `startSpanFast` for optimized path when disabled

dhartunian opened the pull request:
This change attempts to reduce the impact of `StartSpanCtx` when
called with tracing disabled.

A new fast path is introduced via the `spanSpanFast` method which
avoids the more complex logic within `startSpanGeneric`.

Additionally, the regexp-based verbose tracing is modified to use an
atomic pointer instead of a mutex, and the `SpanOption` iterator is
modified to avoid extra copies.

Below are benchmark results which show the most impact on `opts=none` cpu.
```
❯ benchdiff  ./pkg/util/tracing --old ':/consolidate regexp bench' -b -r 'BenchmarkTracer_StartSpanCtx' -d 100000x -c 10
test binaries already exist for a2746cd: tracing: consolidate regexp bench into BenchmarkTr
test binaries already exist for 41fdcda: tracing: add startSpanFast for optimized path when

  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/util/tracing \

name                                                                            old time/op    new time/op    delta
Tracer_StartSpanCtx/opts=none/parallel=true-10                                     164ns ±10%      16ns ±64%  -90.12%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=none/parallel=false-10                                   56.1ns ± 5%    47.7ns ± 5%  -14.96%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                     584ns ± 3%     546ns ±20%   -6.51%  (p=0.002 n=8+9)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                          636ns ± 5%     615ns ± 4%   -3.37%  (p=0.014 n=10+10)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                         318ns ± 3%     308ns ± 3%   -2.92%  (p=0.001 n=10+9)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10       473ns ± 3%     460ns ± 1%   -2.79%  (p=0.000 n=10+8)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10      669ns ± 4%     651ns ± 2%   -2.71%  (p=0.014 n=9+9)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10     353ns ± 2%     348ns ± 4%   -1.32%  (p=0.019 n=9+9)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                    285ns ± 3%     284ns ± 1%     ~     (p=0.923 n=10+9)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                             303ns ± 4%     303ns ± 3%     ~     (p=0.811 n=10+10)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                              590ns ± 2%     585ns ± 3%     ~     (p=0.190 n=9+9)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                       338ns ± 3%     333ns ± 3%     ~     (p=0.052 n=10+10)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                        651ns ± 4%     639ns ± 4%     ~     (p=0.052 n=10+10)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10        910ns ±18%     880ns ±24%     ~     (p=0.436 n=10+10)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                              741ns ± 3%     808ns ± 2%   +9.10%  (p=0.000 n=9+10)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                             500ns ± 1%     722ns ± 1%  +44.49%  (p=0.000 n=8+8)

name                                                                            old alloc/op   new alloc/op   delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                    48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                     48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                    48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                     48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                             48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                              48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                         48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                          48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                       48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                        48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10       96.0B ± 0%     96.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10        96.0B ± 0%     96.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10     48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10      48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                             48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                              51.0B ± 0%     51.0B ± 0%     ~     (all equal)

name                                                                            old allocs/op  new allocs/op  delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                     1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                     1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                              1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                               1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                          1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                           1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                        1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                         1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10        2.00 ± 0%      2.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10         2.00 ± 0%      2.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10       1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                              1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                               1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

Part of: https://github.com/cockroachdb/cockroach/issues/133307
Epic: CRDB-42854

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136931)
<!-- Reviewable:end -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> and thank you!

A tiny bit of advertisement:

You could pull out the benchmark into a first commit as I did here: https://github.com/cockroachdb/cockroach/compare/master...tbg:cockroach:tracing-overhead-improvements?expand=1

then you can do things like this:

```
$ benchdiff  ./pkg/util/tracing --old ':/compare tracing modes' -b -r 'BenchmarkStartSpan/start_span:_1' -d 100000x -c 10
checking out 'c868377'
building benchmark binaries for c868377: tracing: compare tracing modes in BenchmarkStartSp [bazel=true] 1/1 |
checking out 'f52755f'
building benchmark binaries for f52755f: tracing: add `startSpanFast` for optimized path wh [bazel=true] 1/1 \

  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/util/tracing |

name                        old time/op    new time/op    delta
StartSpan/start_span:_1-24    87.2ns ±11%    26.1ns ± 4%  -70.12%  (p=0.000 n=10+10)

name                        old alloc/op   new alloc/op   delta
StartSpan/start_span:_1-24     0.00B          0.00B          ~     (all equal)

name                        old allocs/op  new allocs/op  delta
StartSpan/start_span:_1-24      0.00           0.00          ~     (all equal)
```

Also, there's also `BenchmarkTracer_StartSpanCtx` which seems like a more elaborate version of `BenchmarkStartSpan`. In particular `BenchmarkStartSpan` elides the overhead that arises in practice from wrapping the incoming context. There might be an opportunity for unification here if you want to take it.

The numbers for `BenchmarkTracer_StartSpanCtx` (after prepending a commit to fix that benchmark - see my link above) suggest that this PR could take approx 23% off the CPU usage related to starting spans, but that if we elide the wrapping, we'll see reduction that is similar in magnitude or greater. And this analysis misses the benefit of removing a link from the context value chain in the critical paths, which we have to walk many times over, and ultimately garbage-collecting the allocation.

```
$ benchdiff  ./pkg/util/tracing --old ':/compare tracing modes' -b -r 'BenchmarkTracer_StartSpanCtx/opts=none' -d 100000x -c 30

name                              old time/op    new time/op    delta
Tracer_StartSpanCtx/opts=none-24    97.7ns ± 3%    75.3ns ± 9%  -22.95%  (p=0.000 n=28+29)

name                              old alloc/op   new alloc/op   delta
Tracer_StartSpanCtx/opts=none-24     48.0B ± 0%     48.0B ± 0%     ~     (all equal)

name                              old allocs/op  new allocs/op  delta
Tracer_StartSpanCtx/opts=none-24      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

With noop-spans elided, the overhead in fact melts away to what you observe in your benchmark:

```
$ benchdiff  ./pkg/util/tracing --old ':/compare tracing modes' -b -r 'BenchmarkTracer_StartSpanCtx/opts=none' -d 100000x -c 30
test binaries already exist for 77fa5f6: tracing: compare tracing modes in BenchmarkStartSp
checking out '67a803b'
building benchmark binaries for 67a803b: [experiment] avoid noop span wrapping [bazel=true] 1/1 /

  pkg=1/1 iter=30/30 cockroachdb/cockroach/pkg/util/tracing /

name                              old time/op    new time/op    delta
Tracer_StartSpanCtx/opts=none-24    97.4ns ± 2%    24.5ns ± 2%   -74.80%  (p=0.000 n=27+30)

name                              old alloc/op   new alloc/op   delta
Tracer_StartSpanCtx/opts=none-24     48.0B ± 0%      0.0B       -100.00%  (p=0.000 n=30+30)

name                              old allocs/op  new allocs/op  delta
Tracer_StartSpanCtx/opts=none-24      1.00 ± 0%      0.00       -100.00%  (p=0.000 n=30+30)
```

In other words, there's good evidence that https://github.com/cockroachdb/cockroach/issues/135803 is worth our while.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136931#-:-ODePpnP04nJ9SXo8hfu:b9vupvo)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, and @xinhaoz)

___
*[`pkg/util/tracing/tracer.go` line 1112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136931#-ODeVhcj5QQ2qxDuozz6:-ODeVhcj5QQ2qxDuozz7:bqod1xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b685ba51ee6e88bcc0aa44492c8ecec71a54c78e/pkg/util/tracing/tracer.go#L1112)):*
> ```Go
> 	// `(*spanOptions).recordingType()`. Specifically, we are aiming to
> 	// avoid creating local copies of `opts.Parent` and
> 	// `opts.RemoteParent` hence conditional logic is being inlined and
> ```

could you add a comment in `(*spanOptions).recordingType` as well?


<!-- Sent from Reviewable.io -->

dhartunian committed:
tracing: consolidate regexp bench into BenchmarkTracer_StartSpanCtx

This commit additionally adds a `parallel` true/false run to the
benchmark which runs the span construction using `b.RunParallel`. This
was found to reproduce the CPU attribution to the regexp lookup which
was guarded by a mutex.

Release note: None

dhartunian committed:
tracing: add startSpanFast for optimized path when disabled

This change attempts to reduce the impact of `StartSpanCtx` when
called with tracing disabled.

A new fast path is introduced via the `spanSpanFast` method which
avoids the more complex logic within `startSpanGeneric`.

Additionally, the regexp-based verbose tracing is modified to use an
atomic pointer instead of a mutex, and the `SpanOption` iterator is
modified to avoid extra copies.

Below are benchmark results which show the most impact on `opts=none` cpu.
```
❯ benchdiff  ./pkg/util/tracing --old ':/consolidate regexp bench' -b -r 'BenchmarkTracer_StartSpanCtx' -d 100000x -c 10
test binaries already exist for a2746cd: tracing: consolidate regexp bench into BenchmarkTr
test binaries already exist for 41fdcda: tracing: add startSpanFast for optimized path when

  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/util/tracing \

name                                                                            old time/op    new time/op    delta
Tracer_StartSpanCtx/opts=none/parallel=true-10                                     164ns ±10%      16ns ±64%  -90.12%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=none/parallel=false-10                                   56.1ns ± 5%    47.7ns ± 5%  -14.96%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                     584ns ± 3%     546ns ±20%   -6.51%  (p=0.002 n=8+9)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                          636ns ± 5%     615ns ± 4%   -3.37%  (p=0.014 n=10+10)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                         318ns ± 3%     308ns ± 3%   -2.92%  (p=0.001 n=10+9)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10       473ns ± 3%     460ns ± 1%   -2.79%  (p=0.000 n=10+8)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10      669ns ± 4%     651ns ± 2%   -2.71%  (p=0.014 n=9+9)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10     353ns ± 2%     348ns ± 4%   -1.32%  (p=0.019 n=9+9)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                    285ns ± 3%     284ns ± 1%     ~     (p=0.923 n=10+9)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                             303ns ± 4%     303ns ± 3%     ~     (p=0.811 n=10+10)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                              590ns ± 2%     585ns ± 3%     ~     (p=0.190 n=9+9)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                       338ns ± 3%     333ns ± 3%     ~     (p=0.052 n=10+10)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                        651ns ± 4%     639ns ± 4%     ~     (p=0.052 n=10+10)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10        910ns ±18%     880ns ±24%     ~     (p=0.436 n=10+10)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                              741ns ± 3%     808ns ± 2%   +9.10%  (p=0.000 n=9+10)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                             500ns ± 1%     722ns ± 1%  +44.49%  (p=0.000 n=8+8)

name                                                                            old alloc/op   new alloc/op   delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                    48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                     48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                    48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                     48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                             48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                              48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                         48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                          48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                       48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                        48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10       96.0B ± 0%     96.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10        96.0B ± 0%     96.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10     48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10      48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                             48.0B ± 0%     48.0B ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                              51.0B ± 0%     51.0B ± 0%     ~     (all equal)

name                                                                            old allocs/op  new allocs/op  delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                     1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                     1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                              1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                               1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                          1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                           1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                        1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                         1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10        2.00 ± 0%      2.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10         2.00 ± 0%      2.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10       1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                              1.00 ± 0%      1.00 ± 0%     ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                               1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

Part of: #133307
Epic: CRDB-42854

Release note: None

dhartunian commented:
@tbg thanks for all the detailed help. PR updated with all your suggestions. I added the option to run the all the benchmarks using `b.RunParallel` and did notice that the impact on a parallel benchmark is significantly larger than on the standard one. I don't fully understand why. Do you think this is a useful signal or a red herring?

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thank you!

> I added the option to run the all the benchmarks using `b.RunParallel` and did notice that the impact on a parallel benchmark is significantly larger than on the standard one.

I'm not sure. There's no mutex acquisition in the path that saw the 90%+ reduction, right? Maybe you've optimized memory accesses down to the point where everything now  happens in the L1 cache. I'm not sure, honestly. The easiest explanation is that there was a mutex acquisition that you removed.

> Reviewed 3 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136931#-:-ODl2SbT8m9AbdQ-raaI:bq80jaq)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, and @xinhaoz)



<!-- Sent from Reviewable.io -->

dhartunian commented:
TFTR!

bors r=tbg

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136931)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`pkg/util/tracing/tracer.go` line 1101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136931#-ODlXKgOB7veirCnCGp2:-ODlXKgP21kMgvGcoPYj:b-jh2buy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac4172269b319b602af4206bb19df48e3b19d516/pkg/util/tracing/tracer.go#L1101)):*
> ```Go
> }
> 
> func (t *Tracer) startSpanFast(
> ```

[nit] a comment here explaining how this is different than startSpanGeneric and when it should be used would it be useful.
___
*[`pkg/util/tracing/tracer.go` line 1102 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136931#-ODlWjisAdMNhHM02uXE:-ODlWjisAdMNhHM02uXF:b6bvwmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac4172269b319b602af4206bb19df48e3b19d516/pkg/util/tracing/tracer.go#L1102)):*
> ```Go
> 
> func (t *Tracer) startSpanFast(
> 	ctx context.Context, opName string, opts spanOptions,
> ```

Unrelated drive-by but `spanOptions` is pretty big and I saw the copying of it show up in profiles, we should switch to `opts *spanOptions` (the pointer won't escape).


<!-- Sent from Reviewable.io -->
